## 1. Loading Package 

In [50]:
#Basic import
import pandas as pd
import numpy as np
from datetime import timedelta
import pickle

#Machine Learning
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neighbors import NearestNeighbors
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import classification_report, accuracy_score

## 2. Data Loading

In [51]:
#Import data_mnist
data = pd.read_csv (r'/Users/joan/Desktop/Spring2022/Machine Learning 6364/HW1/data_mnist.csv')

#Import test_mnist
test= pd.read_csv(r'/Users/joan/Desktop/Spring2022/Machine Learning 6364/HW1/test_mnist.csv')

###### Check the data import successfully or not

In [52]:
data.head(5)
#.head(means from the index0 to index5)
#.tail(means the last number)

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [53]:
test.head(10)

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   
5       0       0       0       0       0       0       0       0       0   
6       0       0       0       0       0       0       0       0       0   
7       0       0       0       0       0       0       0       0       0   
8       0       0       0       0       0       0       0       0       0   
9       0       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   
5       0  ...         0         0         0         0         0         0   
6       0  ...         0         0         0         0         0         0   
7       0  ...         0         0         0         0         0         0   
8       0  ...         0         0         0         0         0         0   
9       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  
5         0         0         0         0  
6         0         0         0         0  
7         0         0         0         0  
8         0         0         0         0  
9         0         0         0         0  

[10 rows x 784 columns]

###### 784 columns because doesn't contain label

## 3. Data Convert and Preprocess 

###### 3.1 Convert the dataframe into array (matrix)

In [54]:
data_matrix=np.array(data)

###### 3.2 Split the data into X(features) and Y(target)

In [55]:
#matrix name[row, column]
X=data_matrix[:,1:]
Y=data_matrix[:,0]

###### 3.3 Split data into two sets: (1) Train data set (2) Test data set  {Train-Test split algorithms}

In [56]:
#Both of theses two sets are from data_mnist, and Test data set is use for "evaluating" the model's accuracy.
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.25,random_state=42, stratify=Y)

###### 3.4 Data Preprocessing- StandardScaler

In [57]:
sc_X = StandardScaler()

sc_X.fit(X_train)
X_train_std=sc_X.transform(X_train)
X_test_std=sc_X.transform(X_test)

After preprocessing, we are going to find the best K values to use the model to predict the test file.

## 4. Model- Find the Best K Values

We will going to test the K values from 2 to 5 to see which values have the higher accuracy and store that model.

##### When n_neighbor=2

In [58]:
#perform training
#creating the classifier object
clf_2=KNeighborsClassifier(n_neighbors=2)

#performing training
clf_2.fit(X_train_std, Y_train)

#predict X_test's target
Y_pred=clf_2.predict(X_test_std)

In [59]:
#Classification Report
print("Classification Report:")
print(classification_report(Y_test, Y_pred))

#Acuracy
print('When K=2,the acuracy is', accuracy_score(Y_test, Y_pred)*100)

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.99      0.96      1033
           1       0.93      1.00      0.96      1171
           2       0.91      0.94      0.93      1044
           3       0.89      0.94      0.91      1088
           4       0.90      0.94      0.92      1018
           5       0.90      0.89      0.90       949
           6       0.97      0.95      0.96      1034
           7       0.91      0.93      0.92      1100
           8       0.98      0.82      0.89      1016
           9       0.94      0.85      0.89      1047

    accuracy                           0.93     10500
   macro avg       0.93      0.92      0.92     10500
weighted avg       0.93      0.93      0.92     10500

When K=2,the acuracy is 92.54285714285714


##### When n_neighbor=3

In [60]:
#Perform training
#Creating the classifier object
clf_3=KNeighborsClassifier(n_neighbors=3,weights="distance")

#performing training
clf_3.fit(X_train_std, Y_train)

#predict X_test's target
Y_pred=clf_3.predict(X_test_std)

In [61]:
#Classification Report
print("Classification Report:")
print(classification_report(Y_test, Y_pred))

#Acuracy
print('When K=3,the acuracy is', accuracy_score(Y_test, Y_pred)*100)

Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.98      0.97      1033
           1       0.95      0.99      0.97      1171
           2       0.95      0.94      0.94      1044
           3       0.91      0.93      0.92      1088
           4       0.95      0.92      0.93      1018
           5       0.92      0.91      0.92       949
           6       0.95      0.98      0.96      1034
           7       0.93      0.93      0.93      1100
           8       0.97      0.87      0.92      1016
           9       0.90      0.93      0.92      1047

    accuracy                           0.94     10500
   macro avg       0.94      0.94      0.94     10500
weighted avg       0.94      0.94      0.94     10500

When K=3,the acuracy is 93.93333333333334


##### When n_neighbor=4

In [62]:
#perform training
#creating the classifier object
clf_4=KNeighborsClassifier(n_neighbors=4)

#performing training
clf_4.fit(X_train_std, Y_train)

#predict X_test's target
Y_pred=clf_4.predict(X_test_std)

In [63]:
#Classification Report
print("Classification Report:")
print(classification_report(Y_test, Y_pred))

#Acuracy
print('When K=4,the acuracy is', accuracy_score(Y_test, Y_pred)*100)

Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.98      0.96      1033
           1       0.94      0.99      0.97      1171
           2       0.93      0.94      0.94      1044
           3       0.92      0.93      0.93      1088
           4       0.93      0.94      0.93      1018
           5       0.91      0.90      0.91       949
           6       0.96      0.96      0.96      1034
           7       0.92      0.94      0.93      1100
           8       0.98      0.85      0.91      1016
           9       0.92      0.90      0.91      1047

    accuracy                           0.94     10500
   macro avg       0.94      0.93      0.93     10500
weighted avg       0.94      0.94      0.93     10500

When K=4,the acuracy is 93.53333333333333


##### When n_neighbor=5

In [64]:
#perform training
#creating the classifier object
clf_5=KNeighborsClassifier(n_neighbors=5)

#performing training
clf_5.fit(X_train_std, Y_train)

#predict X_test's target
Y_pred=clf_5.predict(X_test_std)

In [65]:
#Classification Report
print("Classification Report:")
print(classification_report(Y_test, Y_pred))

#Acuracy
print('When K=5,the acuracy is', accuracy_score(Y_test, Y_pred)*100)

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.98      0.96      1033
           1       0.94      0.99      0.96      1171
           2       0.94      0.93      0.94      1044
           3       0.92      0.93      0.93      1088
           4       0.94      0.92      0.93      1018
           5       0.91      0.91      0.91       949
           6       0.96      0.97      0.96      1034
           7       0.93      0.93      0.93      1100
           8       0.97      0.86      0.91      1016
           9       0.90      0.92      0.91      1047

    accuracy                           0.94     10500
   macro avg       0.94      0.93      0.93     10500
weighted avg       0.94      0.94      0.94     10500

When K=5,the acuracy is 93.58095238095238


After testing the K value from 2 to 5, we can tell that when k=3 has the higher accuracy which equals to 93.93%. Therefore, we will store the model when k=3 to predict the test file.

###### Store the Model when K=3

In [66]:
pickle.dump(clf_3, open("Best_model", "wb"))

## 5. Predict - Test.csv

###### 5.1 Load the Best_model

In [67]:
KNN_Model=pickle.load(open("Best_model", 'rb'))

###### 5.2 Convert the dataframe into array

In [68]:
test_matrix=np.array(test)

###### 5.3 Predict 

In [69]:
test_pred=KNN_Model.predict(test_matrix)

###### 5.4 Convert the array into dataframe

In [70]:
pred_DF= pd.DataFrame({'ImageId':np.arange(len(test)),'Label':test_pred})

In [71]:
#Different way to visual: Insert the Label Column into the Test file
test['Label'] = test_pred
#print (test.head(5))

## 6. Submission

In [72]:
print("The Final Result for the test file prediction:")
print(pred_DF)

The Final Result for the test file prediction:
      ImageId  Label
0           0      7
1           1      2
2           2      1
3           3      0
4           4      4
...       ...    ...
9995     9995      2
9996     9996      3
9997     9997      8
9998     9998      5
9999     9999      6

[10000 rows x 2 columns]


In [73]:
pred_DF.to_csv('hw1_submission.csv', index=False)